### PR TITLE
Add stack index to `Node`

### PR DESCRIPTION
**Pull request by ickshonpe**
# Objective

If we add the stack index to `Node` then we don't need to walk the `UiStack` repeatedly during extraction.

## Solution

Add a field `stack_index`  to `Node`.
Update it in `ui_stack_system`.
Iterate queries directly in the UI's extraction systems.

### Benchmarks 
```
cargo run --profile stress-test --features trace_tracy --example many_buttons -- --no-text --no-borders
```

frames (yellow this PR, red main):

frames-per-second

`ui_stack_system`:
ui-stack-system

extract schedule:
extract-schedule

---

## Changelog

* Added the field `stack_index` to `Node`.
* `ui_stack_system` updates `Node::stack_index` after a new `UiStack` is generated.
* The UI's extraction functions iterate a query directly rather than walking the `UiStack` and doing lookups.

### CODE REVIEW

**Comment by ickshonpe** [comment]
> I'm on board with this, but could you label which graph is from which scenario? It's hard to follow.

Yep I ran the benchmarks with the wrong settings for this initially as well. These ones should be a lot clearer.

**Review by alice-i-cecile — APPROVED** [review]

File: crates/bevy_ui/src/stack.rs [hunk 1]
```diff
@@ -55,6 +56,14 @@ pub fn ui_stack_system(
...
     ui_stack.uinodes.reserve(total_entry_count);
     fill_stack_recursively(&mut ui_stack.uinodes, &mut global_context);
+
+    for (i, entity) in ui_stack.uinodes.iter().enumerate() {
+        update_query
+            .get_mut(*entity)
+            .unwrap()
+            .bypass_change_detection()
+            .stack_index = i as u32;
+    }
```
Comment (kristoff3r) [inline crates/bevy_ui/src/stack.rs:66]: Instead of having an extra update query you could have `fill_stack_recursively` set the index, e.g. by taking a `&mut u32` that it uses and increments.

Comment (kristoff3r) [reply crates/bevy_ui/src/stack.rs:66]: Ah it doesn't have the node. I don't know enough about this system to know if that's easy to get on one of the other queries, otherwise this is probably the simplest solution.

Comment (ickshonpe) [reply crates/bevy_ui/src/stack.rs:66]: Yep there might be more efficient ways of updating the `Node` components. I didn't want to make anything more than minor changes with this PR though. The improvement in extraction performance is good enough on its own.

**Review by TimJentzsch — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
@ickshonpe can you resolve comments? I'll merge this ASAP after.

**Comment by TimJentzsch** [comment]
Sorry, I think my other PR caused merge conflicts for this.
Should be easy to resolve though, you just have to replace instances of `color.0.a() == 0.0` by `color.0.is_fully_transparent()`.

**Comment by ickshonpe** [comment]
> @ickshonpe can you resolve comments? I'll merge this ASAP after.

Should be ready now
